client/checkout/classic/upe-split.js

```diff
@@ -329,6 +329,18 @@ jQuery( function ( $ ) {
 		gatewayUPEComponents[ paymentMethodType ].upeElement = upeElement;
 	};
 
+	function maybeToggleAffirm() {
+		const billingCountry = $( '#billing_country' ).val();
+		const affirmContainer = $(
+			'#payment_method_woocommerce_payments_affirm'
+		).parent();
+		if ( 'US' === billingCountry || 'CA' === billingCountry ) {
+			affirmContainer.show();
+		} else {
+			affirmContainer.hide();
+		}
```

Comment:
I'm having some issues enabling Affirm as a payment method, but thought I'd leave a comment in the meantime while I flail around in search for the right migration to run on my local Payments Server instance to end the pain. I'm pretty sure this will work and I don't have too many issues with the implementation. However, it seems like we could implement this in a way that is perhaps slightly more robust and extensible. 

Firstly, it seems like these countries should be both stored somewhere as constants instead of living alone inside this function. For example, we declare [compatible currencies](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-affirm-payment-method.php#L31) or whether the payment method [supports saved payments](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-affirm-payment-method.php#L29) as properties of the relevant `UPE_Payment_Method` class. These are then passed to the client via the [`paymentMethodsConfig` global parameter](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-upe-checkout.php#L152), whose contents are filled [here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-upe-checkout.php#L229). 

This is one potential approach, but alternatively--since these values currently have no utility on the back-end of the plugin (unless we decide to toy with geolocation or anything like that)--perhaps these settings could live somewhere in a client-side constant somewhere like [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/constants.js)? This would basically make it easier to mirror similar functionality in the future, should we need to enable a new payment method with similar geographical restrictions.

Secondly, this implementation will successfully hide Affirm on the shortcode checkout for the split UPE. We will soon need to ensure that this integration also supports the blocks checkout and pay for order page, as well as all three checkouts when the deferred intent UPE is enabled. This definitely seems like it could be better placed inside a utility function ([here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/utils/upe.js) or [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/utils/checkout.js) maybe), so that we wouldn't have to duplicate client-side code in the plugin.

Finally, concerning hiding the gateways at checkout, I just want to confirm that this is simply to prevent an error being returned by Stripe. Obviously, since the gateways are still present in the DOM, an end-user could still revive these hidden payment methods and attempt to use them to complete a transaction. Hopefully, in this case, they would only be met by an error and nothing else harmful or unintended could happen from our perspective. 

Sorry, this is quite a large comment with some pretty scattered feedback, so hopefully it's not unwelcome! If there is an urgency to add this integration, many of the above suggestions and refactors could easily be handled in separate issues, so let me know if you'd prefer to proceed in that way instead of handling everything here. Also, it's likely I've made many miscomprehensions along the way--especially since I haven't actually been able to run the code yet 😅--so please don't hesitate to correct any incorrect assertions I may have made along the way and apologies in advance for the holes in my swiss-cheese understanding. Cheers!

Reply:
> Sorry, this is quite a large comment with some pretty scattered feedback, so hopefully it's not unwelcome! 

Your feedback is most welcomed here. Is overlaps with most of my own thoughts giving me a good green light for further improvements. 

> Finally, concerning hiding the gateways at checkout, I just want to confirm that this is simply to prevent an error being returned by Stripe.[..]Hopefully, in this case, they would only be met by an error and nothing else harmful or unintended could happen from our perspective.

Exactly, customer can not in any way tamper checkout, only expose them selves to flow that will always fail.